## Homework 3
This homework asks you to scrape from three different sources, please follow the instructions and do the best you can. With the exception of the first cell that imports the URL request and the beautiful soup library, I have not pre-written any code for you. Do not be afraid--you may look at the tutorial for examples, as well as the Beautiful Soup documentation, and any other Python resource (such a Stack overflow) if need be. 

This homework is intended to be challenging, if you only get 70% of the stuff done than you did very well.

In [1]:
from urllib.request import urlopen
from bs4 import BeautifulSoup

## Shakespeare 
I've posted some relatively simple HTML of the first act of William Shakespeare's The Tempest here: http://floatingmedia.com/columbia/tempest.html please use that URL to download the HTML and put it through Beautiful Soup. (It also may be very helpful for you to go to that page in chrome and inspect elements why you do this.) 
**Please note: there are 2 scenes in Act 1**



In the cells below you should you make two lines--a variable that requests and reads the HTML from the URL (http://floatingmedia.com/columbia/tempest.html), and a second variable that passes the raw HTML into Beautiful Soup. You will use that beautiful soup variable to search the HTML.

In [2]:
my_url = "http://floatingmedia.com/columbia/tempest.html"
raw_html = urlopen(my_url).read() 
soup_doc = BeautifulSoup(raw_html, "html.parser")  
#print(soup_doc.prettify()) #--> this will give you the HTML of the entire document 


Get the title of the play:

In [3]:
soup_doc.title.string 

'The Tempest'

Get the HTML that contains the setting of Act One Scene 1:

In [4]:
#print(soup_doc.prettify())
setting = soup_doc.find('p') 
setting

<p class="playsubhead">Act I, Scene 1</p>

Get the setting of Act One Scene 1 (as a string):

In [5]:
setting.string 

'Act I, Scene 1'

Get the setting of scene 2

In [6]:
settings = soup_doc.find_all('div')
setting_scene_2 = settings[2].find('strong') 
setting_scene_2 

<strong>The island. Before PROSPERO’S cell.</strong>

Get the name of the first character in Scene 1

In [7]:
name = soup_doc.find_all('li') 
name 
first_name = name[0].find('strong') 
first_name.string 


'Master. '

Get a list of all of characters in Scene 1 (repeats are fine, you need a loop here)

In [8]:
scene_1 = soup_doc.find_all('div')[1]
all_characters = scene_1.find_all('li')
for people in all_characters:
    print(people.find('strong').string) 

Master. 
Boatswain. 
Master. 
Boatswain. 
Alonso. 
Boatswain. 
Antonio. 
Boatswain. 
Gonzalo. 
Boatswain. 
Gonzalo. 
Boatswain. 
Gonzalo. 
Boatswain. 
Sebastian. 
Boatswain. 
Antonio. 
Gonzalo. 
Boatswain. 
Mariners. 
Boatswain. 
Gonzalo. 
Sebastian. 
Antonio. 
Gonzalo. 
Antonio. 
Sebastian. 
Gonzalo. 


**Bonus:** write a script that takes all the duplicates out of the list. (This is a python thing, not a beautiful soup thing.)

In [9]:
### 

Display every stage direction in _scene 2_

In [10]:
stage_directions = soup_doc.find_all('div')
directions2 = stage_directions[3].find_all('p') 
directions2

[<p class="stagedir">[Enter PROSPERO and MIRANDA]</p>,
 <p class="stagedir">[Enter ARIEL]</p>,
 <p class="stagedir">[Exit]</p>,
 <p class="stagedir">[Enter CALIBAN]</p>,
 <p class="stagedir">[Draws, and is charmed from moving]</p>,
 <p class="stagedir">[Exeunt]</p>]

Get the HTML containing Miranda's first speech in Scene 2

In [11]:
miranda = stage_directions[3].find('ul')
miranda

<ul><li class="playtext"><strong>Miranda. </strong>If by your art, my dearest father, you have
 <span class="playlinenum">85</span><br/>Put the wild waters in this roar, allay them.
<br/>The sky, it seems, would pour down stinking pitch,
<br/>But that the sea, mounting to the welkin's cheek,
<br/>Dashes the fire out. O, I have suffered
<br/>With those that I saw suffer: a brave vessel,
 <span class="playlinenum">90</span><br/>Who had, no doubt, some noble creature in her,
<br/>Dash'd all to pieces. O, the cry did knock
<br/>Against my very heart. Poor souls, they perish'd.
<br/>Had I been any god of power, I would
<br/>Have sunk the sea within the earth or ere
 <span class="playlinenum">95</span><br/>It should the good ship so have swallow'd and
<br/>The fraughting souls within her.
</li></ul>

Now display those same lines without the speaker's name (need to use siblings and a loop to get this)

In [12]:
#miranda.find('strong').parent
new_speech = miranda.find('strong')
just_speech = new_speech.find_next_siblings()
just_speech

[<span class="playlinenum">85</span>,
 <br/>,
 <br/>,
 <br/>,
 <br/>,
 <br/>,
 <span class="playlinenum">90</span>,
 <br/>,
 <br/>,
 <br/>,
 <br/>,
 <br/>,
 <span class="playlinenum">95</span>,
 <br/>,
 <br/>]

In [13]:
for lines in just_speech:
    print(lines.text)

85





90





95




Get the HTML containing the speech after Miranda's

## Box Office Mojo 
I've posted some slightly simplified HTML of the current worldwide box office leaders here: http://floatingmedia.com/columbia/boxoffice.htm please use that URL to download the HTML and put it through Beautiful Soup. (It also may be very helpful for you to go to that page in chrome and inspect elements why you do this.) 

For this part of the assignment you should focus mostly on tr (table rows) amd td (table cells) -- and search them using find_all('tr')[1] in order to get a specific row, and find_all('td')[2] to get a specific cell.

Just like with the Shakespeare write two lines get the raw HTML and then transform the raw HTML into a beautiful soup variable

In [14]:
#http://floatingmedia.com/columbia/boxoffice.htm
my_url = "http://floatingmedia.com/columbia/boxoffice.htm"
raw_html = urlopen(my_url).read() 
soup_doc = BeautifulSoup(raw_html, "html.parser") 




Get the text "2017 worldwide grosses"

In [15]:
soup_doc.find('h1').string

'2017 WORLDWIDE GROSSES'

Get an HTML list of all of the column names from the header of the table: rank, title, etc...

In [16]:
getting_columns = soup_doc.find_all('table') 
getting_columns
tr_columns = getting_columns[0].find_all('tr')
tr_columns
td_columns = tr_columns[0].find_all('td') 
td_columns

[<td align="center">Rank</td>,
 <td align="center">Title</td>,
 <td align="center">Studio</td>,
 <td align="center">Worldwide</td>,
 <td align="center">Domestic</td>,
 <td align="center">Overseas</td>]

Make a variable called "headers" and set it equal to the list of strings of the column names and header. In the next line write: "headers[1]" your output should be "Title"

In [17]:
headers = td_columns
headers[1]

<td align="center">Title</td>

Get the HTML for the row that contains the #1 movie

In [18]:
getting_columns = soup_doc.find_all('table') 
getting_columns
tr_columns = getting_columns[0].find_all('tr')
tr_columns
td_columns2 = tr_columns[1].find_all('td') 
td_columns2

[<td align="center"><font size="2">1</font></td>,
 <td><font size="2"><a href="http://www.boxofficemojo.com/movies/?id=beautyandthebeast2017.htm"><b>Beauty and the Beast (2017)</b></a></font></td>,
 <td><font size="2">BV</font></td>,
 <td align="right"><font size="2"><b>$1,241.3</b></font></td>,
 <td align="right"><font size="2">$502.2</font></td>,
 <td align="right"><font size="2">$739.1</font></td>]

Get the HTML for the row that contains the #10 movie 

In [19]:
td_columns10 = tr_columns[10].find_all('td')
td_columns10

[<td align="center"><font size="2">10</font></td>,
 <td><font size="2"><a href="http://www.boxofficemojo.com/movies/?id=xxx3.htm"><b>xXx: The Return of Xander Cage</b></a></font></td>,
 <td><font size="2">Par.</font></td>,
 <td align="right"><font size="2"><b>$346.1</b></font></td>,
 <td align="right"><font size="2">$44.9</font></td>,
 <td align="right"><font size="2">$301.2</font></td>]

Get a list of the top 10 films (You'll need to use the loop and specify a range for the loop.) The list should look like this:

`Beauty and the Beast (2017)
The Fate of the Furious
Guardians of the Galaxy Vol. 2
Logan
Kong: Skull Island
Pirates of the Caribbean: Dead Men Tell No Tales
The Boss Baby
Fifty Shades Darker
Your Name.
xXx: The Return of Xander Cage`

In [20]:
#How do you specify a range? head() won't work with this 
getting_columns = soup_doc.find_all('table') 
getting_columns
tr_columns = getting_columns[0].find_all('tr')
tr_columns

for movies in tr_columns:
    print(movies.find('b'))
    

None
<b>Beauty and the Beast (2017)</b>
<b>The Fate of the Furious</b>
<b>Guardians of the Galaxy Vol. 2</b>
<b>Logan</b>
<b>Kong: Skull Island</b>
<b>Pirates of the Caribbean: Dead Men Tell No Tales</b>
<b>The Boss Baby</b>
<b>Fifty Shades Darker</b>
<b>Your Name.</b>
<b>xXx: The Return of Xander Cage</b>
<b>The Great Wall</b>
<b>Resident Evil: The Final Chapter</b>
<b>The LEGO Batman Movie</b>
<b>Split</b>
<b>Kung Fu Yoga</b>
<b>Get Out</b>
<b>Journey to the West: The Demons Strike Back</b>
<b>Wonder Woman</b>
<b>A Dog's Purpose</b>
<b>Smurfs: The Lost Village</b>
<b>Alien: Covenant</b>
<b>Ghost in the Shell (2017)</b>
<b>John Wick: Chapter Two</b>
<b>Duckweed</b>
<b>Power Rangers (2017)</b>
<b>King Arthur: Legend of the Sword</b>
<b>Buddies in India</b>
<b>Railroad Tigers</b>
<b>Life (2017)</b>
<b>Some Like It Hot (Qing Shung)</b>
<b>The Shack</b>
<b>Rings</b>
<b>Underworld: Blood Wars</b>
<b>Going in Style (2017)</b>
<b>Baywatch</b>
<b>One Piece Film: Gold</b>
<b>Monster Trucks</b>

Display as strings the top 10 films by name and worldwide gross. Like this:

`Beauty and the Beast (2017) $502.2
The Fate of the Furious $223.9
Guardians of the Galaxy Vol. 2 $356.7
Logan $226.2
Kong: Skull Island $167.8
Pirates of the Caribbean: Dead Men Tell No Tales $117.6
The Boss Baby $171.0
Fifty Shades Darker $114.4
Your Name. $4.9
xXx: The Return of Xander Cage $44.9`

In [21]:
rows = soup_doc.find_all('tr')
rows

for row in rows:
    bs = row.find_all('b')
    if len(bs) == 2: 
        print(bs[1].text)
        print(bs[0].text) 
    print('---') 

---
$1,241.3
Beauty and the Beast (2017)
---
$1,229.4
The Fate of the Furious
---
$818.6
Guardians of the Galaxy Vol. 2
---
$612.4
Logan
---
$565.9
Kong: Skull Island
---
$510.5
Pirates of the Caribbean: Dead Men Tell No Tales
---
$483.6
The Boss Baby
---
$378.8
Fifty Shades Darker
---
$354.4
Your Name.
---
$346.1
xXx: The Return of Xander Cage
---
$332.0
The Great Wall
---
$312.2
Resident Evil: The Final Chapter
---
$310.8
The LEGO Batman Movie
---
$276.9
Split
---
$254.2
Kung Fu Yoga
---
$247.5
Get Out
---
$246.6
Journey to the West: The Demons Strike Back
---
$240.0
Wonder Woman
---
$193.1
A Dog's Purpose
---
$189.6
Smurfs: The Lost Village
---
$174.6
Alien: Covenant
---
$169.8
Ghost in the Shell (2017)
---
$166.8
John Wick: Chapter Two
---
$151.1
Duckweed
---
$140.2
Power Rangers (2017)
---
$130.4
King Arthur: Legend of the Sword
---
$109.6
Buddies in India
---
$102.2
Railroad Tigers
---
$98.9
Life (2017)
---
$95.7
Some Like It Hot (Qing Shung)
---
$93.5
The Shack
---
$83.1
Rings
-

Display the link for the movie `Logan`:

In [22]:
td_columns4 = tr_columns[4].find_all('td')
td_columns4

for logan in td_columns4:
    #print(logan.find('a href'))   
    print(logan)



<td align="center"><font size="2">4</font></td>
<td><font size="2"><a href="http://www.boxofficemojo.com/movies/?id=wolverine2017.htm"><b>Logan</b></a></font></td>
<td><font size="2">Fox</font></td>
<td align="right"><font size="2"><b>$612.4</b></font></td>
<td align="right"><font size="2">$226.2</font></td>
<td align="right"><font size="2">$386.2</font></td>


In [23]:
#I'm having difficulty making the code work before to just pull out the <a> 
#link = tr_columns.find('a')  
#logan_url = link['href']
#print(logan_url) 

Create a list for `Logan` that looks like this:

`['4', 'Logan', 'Fox', '$612.4', '$226.2', '$386.2']`

In [24]:
for logan in td_columns4:
    print(logan.string)

4
Logan
Fox
$612.4
$226.2
$386.2


Here is some premade code. If you got your headers list and your logan_list right, this will make you a nice dictionary for Logan.

In [25]:
logan_dict = dict(zip(headers, logan_list))
print(logan_dict)

NameError: name 'logan_list' is not defined

## Advanced Challenge
Pitchfork is no go, so try to get out all of the artists and albums (and synopses) from NPRs 2015 list: http://www.npr.org/2015/12/07/458447213/npr-musics-50-favorite-albums-of-2015

Use as many cells as you would like, and don't forget to make a new URL request and a new beautiful soup variable